Extended `balance_root` to handle root underflow case

krishvishal opened the pull request:
Balance root now handles root underflow case:

- Root becomes underflow when its empty. Its a special case.
- We try and copy root's child into root.
- If successful we delete the child and update pointers.
- If unsuccessful because the child can't fit inside the root, we early return and handle the child overflow in `balance_leaf`

krishvishal commented:
I'm not sure why the test is failing only on windows. I have fixed the same error on my ubuntu machine. This is the only check that's failing.

jussisaurio commented on core/storage/btree.rs:
```diff
@@ -1191,83 +1198,191 @@ impl BTreeCursor {
+        // In SQLite, we want to leave some space for future cell pointers
+        // Each cell pointer takes 2 bytes
+        const CELL_POINTER_SIZE: usize = 2;
+        const ESTIMATED_FUTURE_CELLS: usize = 4; // Leave room for a few more cells
```
Does this come from somewhere, e.g. sqlite source code? If so, good to leave a link as comment

krishvishal replied:
I've removed this function to keep this PR close handling just the root and for checking root overflow, we just need to check if there the page contains any overflow cells. SQLite here: https://github.com/sqlite/sqlite/blob/c2e400af042bdd7d21e159a41fcf34c05398044c/src/btree.c#L9059-L9060

jussisaurio commented on core/storage/btree.rs:
```diff
@@ -1191,83 +1198,191 @@ impl BTreeCursor {
+        }
+    }
+
+    // Helper function to handle page 1's special case
```
This should at least describe what the special case is and why it occurs

jussisaurio replied:
I would rename the function to hint at what the special case is

krishvishal replied:
This is not mentioned in SQLite. The overflow case is almost copied from previous `balance_root` implementation which handled only overflow case.

`balance_deeper` in SQLite handles the root overflow case. But it works a bit differently. It allocates a new child page and makes it right-child page and then it copies the contents into the child page. Then `balance_nonroot` will handle the child page overflow by splitting the page.

You can find `balance_deeper` docs and implementation here: https://github.com/sqlite/sqlite/blob/master/src/btree.c#L8940-L9004

krishvishal commented:
@pereman2 I've looked at both the references below and thought it is better to handle root underflow case in the `balance_root` itself.

I think the comment from [1] describes root underflow case.

```
    /* The root page of the b-tree now contains no cells. The only sibling
    ** page is the right-child of the parent. Copy the contents of the
    ** child page into the parent, decreasing the overall height of the
    ** b-tree structure by one. This is described as the "balance-shallower"
    ** sub-algorithm in some documentation.
    **
    ** If this is an auto-vacuum database, the call to copyNodeContent()
    ** sets all pointer-map entries corresponding to database image pages
    ** for which the pointer is stored within the content being copied.
    **
    ** It is critical that the child page be defragmented before being
    ** copied into the parent, because if the parent is page 1 then it will
    ** by smaller than the child due to the database header, and so all the
    ** free space needs to be up front.


```

[1] https://github.com/sqlite/sqlite/blob/master/src/btree.c#L8867-L8881
[2]https://www.sqlite.org/btreemodule.html#:~:text=4.2.5.2%20Balance%20Shallower

pereman2 commented on core/storage/btree.rs:
```diff
@@ -1191,83 +1198,191 @@ impl BTreeCursor {
+                    self.usable_space(),
+                );
+                let buf = child_contents.as_ptr();
+                cells.push(buf[start..start + len].to_vec());
```
Can't we `insert_into_cell` here instead of cloning vecs?

pereman2 commented on core/storage/btree.rs:
```diff
@@ -1191,83 +1198,191 @@ impl BTreeCursor {
-                let new_root_buf = new_root_page_contents.as_ptr();
-                new_root_buf[0..DATABASE_HEADER_SIZE]
-                    .copy_from_slice(&current_root_buf[0..DATABASE_HEADER_SIZE]);
+        if self.is_underflow(contents) {
```
maybe do:
```rust
if self.is_overflow(contents) {
...
} else if self.is_underflow(contents) {
...
} else {
unreachable!("balance_root was called where we didn't have any overflow or underflow")
}
```

pereman2 commented on core/storage/btree.rs:
```diff
@@ -1191,83 +1198,191 @@ impl BTreeCursor {
-                    .copy_from_slice(&current_root_buf[0..DATABASE_HEADER_SIZE]);
+        if self.is_underflow(contents) {
+            let child_page_id = contents.rightmost_pointer().unwrap();
+            let child_page = self
+                .pager
+                .read_page(child_page_id as usize)
+                .expect("This shouldn't have happened, child page not found");
```
Better to propagate errors
```suggestion
            let child_page = self
                .pager
                .read_page(child_page_id as usize)?;
```

And I think we are not handling the case where child page is not loaded/locked?

pereman2 commented on core/storage/btree.rs:
```diff
@@ -1191,83 +1198,191 @@ impl BTreeCursor {
+                return;
+            }
+
+            let grandchild_ptr = child_contents.rightmost_pointer();
```
rename `grandchild_ptr` -> `child_rightmost_pointer`? There could be a lot of grand childs so it this feels more exact.

pereman2 replied:
Moreover, if we are balancing root because it doesn't have any more cells, why do we expect the child to not be a leaf page? I mean, if you take rightmost pointer it means that child is a interior page, therefore, the tree is not balanced.

pereman2 commented on core/storage/btree.rs:
```diff
@@ -1285,8 +1400,7 @@ impl BTreeCursor {
         let contents = page.get().contents.as_mut().unwrap();
-        let buf = contents.as_ptr();
-        buf.fill(0);
+        contents.write_u32(0, 0); // Initialize next overflow page pointer to 0
```
Why 0, can we use one of the constants? This 0 means `PAGE_HEADER_OFFSET_PAGE_TYPE`, are you sure this is what this 0 write means?

krishvishal replied:
Yes, what I did is wrong. Will fix it.

krishvishal replied:
@pereman2 can you please tell me what's happening here?: https://github.com/sqlite/sqlite/blob/c2e400af042bdd7d21e159a41fcf34c05398044c/src/btree.c#L7166-L7172

This is my understanding:

```c

// Writes the page number of the new overflow page (pgnoOvfl) to pPrior. For the first overflow page, pPrior points to the end of the local cell content.
put4byte(pPrior, pgnoOvfl);

// ignore this
releasePage(pToRelease);
pToRelease = pOvfl;

// Points pPrior to the start of the new overflow page's data area
pPrior = pOvfl->aData;

// Writes 0 at the start of the overflow page, which shows its the last overflow page
put4byte(pPrior, 0);

// Sets pPayload to point past the 4-byte pointer. This is where the actual data will be stored in the overflow page
pPayload = &pOvfl->aData[4];

//Calculates remaining space for data in this overflow page. Usable space minus 4 bytes used for the next-page pointer
spaceLeft = pBt->usableSize - 4;
```

Where are we doing `put4byte(pPrior, 0);` equivalent in limbo?

krishvishal commented:
@pereman2 @jussisaurio I've added the following changes, can you please take a look?

1. Added an unreachable!() to the if-else under/overflow checker.
2. Directly inserted to cells using `insert_into_cell()` instead of cloning vecs into an external vector
3. Propagated errors using `?` instead of `except` and handled the case where child page is not loaded/locked.
4. Restored `allocate_overflow_page()` function to previous state.

krishvishal commented:
I think this is ready for review. 

1. Added `copy_node_content` function which copies a page's along with its overflow cells to dst

2. Added `zero_page` function which zeros's page properly.
3. Fixed the bug on windows which was due to bad offset handling and the `test_sequential_overflow_page` passes now on Windows.
4. Now `balance_root` mirror's closely what sqlite does.

---------------------
I will add tests for root underflow case when delete is functional.

pereman2 commented on core/storage/btree.rs:
```diff
@@ -1191,82 +1198,169 @@ impl BTreeCursor {
+            );
+            total_used += cell_len;
+        }
+        for overflow_cell in &page.overflow_cells {
+            total_used += overflow_cell.payload.len();
+        }
```
Do overflow cells count? I mean, no overflow cell is there inside the page. If there is an overflow cell, regardless of the used space we have to balance.

krishvishal replied:
This calculation is not needed for now and the calculation is wrong. I'm using this function inside in `is_underflow`. I've simplified both `is_underflow` and `is_overflow` to handle only `balance_root` case for. I will extend them in the future.

I will remove this function. This can be resolved.

krishvishal commented on core/storage/btree.rs:
```diff
@@ -1191,82 +1198,169 @@ impl BTreeCursor {
+        Ok(())
+    }
+
+    fn zero_page(&self, page: &PageRef, page_type: PageType, is_leaf: bool) {
+        let contents = page.get().contents.as_mut().unwrap();
```
In my understanding, since `write_u16` and other write functions add page header size offsets when writing data to a location we need not handle page1 specially. Is my understanding correct?

krishvishal commented on core/storage/sqlite3_ondisk.rs:
```diff
@@ -534,11 +538,14 @@ impl PageContent {
-            PageType::TableInterior => Some(self.read_u32(8)),
-            PageType::IndexLeaf => None,
-            PageType::TableLeaf => None,
+        match self.maybe_page_type() {
+            Some(page_type) => match page_type {
+                PageType::IndexInterior => Some(self.read_u32(8)),
+                PageType::TableInterior => Some(self.read_u32(8)),
+                PageType::IndexLeaf => None,
+                PageType::TableLeaf => None,
+            },
+            None => None, // Handle overflow pages by returning None
         }
```
Removing these changes. I have made them for `compute_used_space` which is wrong and not needed for root balancing.

krishvishal commented on core/storage/btree.rs:
```diff
@@ -1191,82 +1191,154 @@ impl BTreeCursor {
+    fn zero_page(&self, page: &PageRef) {
+        let contents = page.get().contents.as_mut().unwrap();
+
+        contents.write_u32(PAGE_HEADER_OFFSET_FIRST_FREEBLOCK, 0);
```
@pereman2 I've written `write_u32` here instead of `write_u16` because SQLite does memset to 4 bytes: https://github.com/sqlite/sqlite/blob/c2e400af042bdd7d21e159a41fcf34c05398044c/src/btree.c#L2277

penberg commented:
This pull request has been marked as stale due to inactivity. It will be closed in 7 days if no further activity occurs.

penberg commented:
This pull request has been closed due to inactivity. Please feel free to reopen it if you have further updates.
